pkg/dockerregistry/server/tagservice.go

```diff
@@ -0,0 +1,56 @@
+package server
+
+import (
+	"github.com/docker/distribution"
+	"github.com/docker/distribution/context"
+	"github.com/docker/distribution/digest"
+
+	imageapi "github.com/openshift/origin/pkg/image/api"
+)
+
+type tagService struct {
+	distribution.TagService
+
+	imageStream *imageapi.ImageStream
+}
+
+func (t tagService) Get(ctx context.Context, tag string) (distribution.Descriptor, error) {
+	context.GetLogger(ctx).Debug("(tagService).Get: starting")
+
+	te := imageapi.LatestTaggedImage(t.imageStream, tag)
+	if te == nil {
+		return distribution.Descriptor{}, distribution.ErrTagUnknown{Tag: tag}
+	}
+	dgst, err := digest.ParseDigest(te.Image)
+	if err != nil {
+		return distribution.Descriptor{}, err
+	}
+
+	return distribution.Descriptor{Digest: dgst}, nil
+}
+
+func (t tagService) All(ctx context.Context) ([]string, error) {
+	context.GetLogger(ctx).Debug("(tagService).All: starting")
+
+	tags := []string{}
+	for tag := range t.imageStream.Status.Tags {
+		tags = append(tags, tag)
+	}
+	return tags, nil
+}
+
+func (t tagService) Lookup(ctx context.Context, digest distribution.Descriptor) ([]string, error) {
+	context.GetLogger(ctx).Debug("(tagService).Lookup: starting")
+	return t.TagService.Lookup(ctx, digest)
+}
+
+// TODO(legion): Rewrite this method to store tag in etcd.
```

Comment:
> We already have tags on a filesystem. If you create a tag using the docker tag, it will be created by the docker/distribution on the filesystem. If you're going to use oc tag, they will be in the etcd.

Really? I don't think so:

    $ oc version
    oc v1.3.0-alpha.0-436-gafc379b-dirty
    kubernetes v1.3.0-alpha.1-331-g0522e63

    $ docker version | grep Version
    Version:         1.10.3
    Version:         1.10.3

    $ docker tag docker.io/hello-world registry.rhel7-ose-registry.vm/imagetags/hello-world
    $ docker push registry.rhel7-ose-registry.vm/imagetags/hello-world
    ...
    latest: digest: sha256:c9b94482f3364819d342785ea9b266d790e04d2d207e345f53ccb703461e23a7 size: 1939
    $ docker tag registry.rhel7-ose-registry.vm/imagetags/hello-world:{latest,v2}
    $ docker push registry.rhel7-ose-registry.vm/imagetags/hello-world:v2
    The push refers to a repository [registry.rhel7-ose-registry.vm/imagetags/hello-world]
    ...
    v2: digest: sha256:94948d7a487b22e7200a5a130b2fd1f8ea4f29809091d0ee5aab68e6a3b21634 size: 1935
    $ tree /var/lib/docker-registry/docker/registry/v2/repositories
    /var/lib/docker-registry/docker/registry/v2/repositories
    └── imagetags
        └── hello-world
            ├── _layers
            │   └── sha256
            │       ├── 4276590986f6a900029d4afb65b0969777f5ff9f70a00f896ec3b7372072381c
            │       │   └── link
            │       ├── 94df4f0ce8a4d4d4c030b9bdfe91fc9cf6a4b7be914542315ef93a046d520614
            │       │   └── link
            │       └── a3ed95caeb02ffe68cdd9fd84406680ae93d633cb16422d00e8a7c22955b46d4
            │           └── link
            ├── _manifests
            │   └── revisions
            │       └── sha256
            │           ├── 94948d7a487b22e7200a5a130b2fd1f8ea4f29809091d0ee5aab68e6a3b21634
            │           │   └── signatures
            │           │       └── sha256
            │           │           └── 60ff1e3dc64c95a4f6bb72afd214c2206b7538a4c23bb7f3faaf77f50bfc5c69
            │           │               └── link
            │           └── c9b94482f3364819d342785ea9b266d790e04d2d207e345f53ccb703461e23a7
            │               └── signatures
            │                   └── sha256
            │                       └── e3be5b44bb12d47e037c2bf554e76de631955b3e7dca773656955e4c50610f83
            │                           └── link
            └── _uploads

    19 directories, 5 files

No trace of tags with current registry image. It looks different when you push to upstream registry (v2.2.1 and some other random image):

    ...
    │   └── ose-docker-registry
    │       ├── _layers
    │       │   └── sha256
    │       │       ├── 315f29c1a1da92001474cb1dc5761a768f73ffe2c33eaebbeabdb6aef2ed6694
    │       │       │   └── link
    │       │       ├── 631d0e21f649c7fd20d09b786453c8d9953e7075dcff9e4c4f587e3013656d14
    │       │       │   └── link
    │       │       └── 704b81d96ecd968d345a04a2c8ef2b7eb16f9dbb00737b70dc6e24777f6de721
    │       │           └── link
    │       ├── _manifests
    │       │   ├── revisions
    │       │   │   └── sha256
    │       │   │       └── 3b07d002890ce7080f4dcf344ea9754ad1d699aaa01c78b51e5305a857c3d248
    │       │   │           ├── link
    │       │   │           └── signatures
    │       │   │               └── sha256
    │       │   │                   └── c517f0131e1c6eb09d8f978ad2e2102a159d178f699ea748e31a67263df8040a
    │       │   │                       └── link
    │       │   └── tags
    │       │       └── latest
    │       │           ├── current
    │       │           │   └── link
    │       │           └── index
    │       │               └── sha256
    │       │                   └── 3b07d002890ce7080f4dcf344ea9754ad1d699aaa01c78b51e5305a857c3d248
    │       │                       └── link
    │       └── _uploads
    ...

You may see the tags present in your setup because of the rebase - which is a **regression**.

> Tags already created by docker needs to be imported into etcd

Tag on registry side is a result of `docker push` and we need to handle it the same way as before.